### ISSUE TITLE
Drone development and directions

### ISSUE THREAD

**Issue by juanibiapina** [drone/drone#1073]
Hi,

We're considering using drone in our organisation. We have a basic setup working integrated with gitlab, and it works great.

Our needs, however, go beyond what drone currently supports. We're thinking of using some of our time to seriously contribute do drone. In order to make a decision, we would like to understand what is the current development plans.

I guess the first question would be: Whats you plan for the 0.4 release? It seems the two branches named after "0.4" have greatly diverged from master, with some big changes regarding builds and the database. Is 0.4 going to be released with all the work from those branches merged into master?

**Comment by bradrydzewski** [comment]
> We're considering using drone in our organisation. We have a basic setup working integrated with gitlab, and it works great.

awesome, glad to hear!

> Our needs, however, go beyond what drone currently supports. We're thinking of using some of our time to seriously contribute do drone. In order to make a decision, we would like to understand what is the current development plans.

I would love to collaborate with you guys to beef up Drone's capabilities

> I guess the first question would be: Whats you plan for the 0.4 release?

0.4 introduces some very major enhancements including matrix builds, a more robust yaml file, plugins and more. The goal with the 0.4 release isn't necessarily stability, since that will take time given the amount of low-level changes that have taken place. The goal with 0.4 is, instead, to release stable API definitions, plugin definitions, yaml definitions, and configuration instructions. Basically, if you start using Drone 0.4, we want to make sure subsequent releases don't "break things". We want people to feel confident writing plugins, integrations and using the new yaml format with fear of things changing.

Subsequent releases (0.4.1, 0.4.N, 0.5, etc) would focus on improving how Drone works under the covers (backend Go code), fixing bugs, improving reliability, adding new plugins, etc.

So this would be a great time to understand what type of changes Drone would require to meet your needs, since breaking compatibility will get more difficult over time.

> Is 0.4 going to be released with all the work from those branches merged into master?

Yes

**Comment by juanibiapina** [comment]
We're mainly interested in these features to start:

1 - Build pipeline. Sequence of jobs triggered by one commit that run in order if the previous one passed.
2 - Manual trigger. Some step of the pipeline would require user confirmation.

I have looked at some of the issues regarding 0.4 and couldn't find anything definitive on build pipelines for the same project.

Also: How usable is the 0.4 branch? Would we be able to upgrade our server from master to 0.4 and keep using it? We're ok with breaking API changes, etc, if they are relatively small. What concerns us most is the code not being in a usable state at any minute, or having to make a major jumps (like master -> 0.4 right now).

**Comment by bradrydzewski** [comment]
> Build pipeline. Sequence of jobs triggered by one commit that run in order if the previous one passed.

This is somewhat supported in 0.4 with the new yaml (see #902). We may need to adjust some of the semantics and ensure ordering (publish, notify, deploy steps are represented as maps and don't guarantee ordering) which won't be too difficult.

It would be interesting to see how far we could get with the 0.4 yaml and where it falls short.

> Manual trigger. Some step of the pipeline would require user confirmation.

I think we could support this in 0.4 today using plugins. I could see a plugin that:

* sends out emails (or whatever) requesting approval
* blocks
* polls for approval
* unblocks once approval is received or fails if disapproved
* build continues

> I have looked at some of the issues regarding 0.4 and couldn't find anything definitive on build pipelines for the same project.

Have a look at #902 and let me know your thoughts.  I also think it would be helpful to understand the pipeline requirement further. Should we build pipelines into Drone? Or should we build a separate pipeline app that tightly integrates with Drone? Drone could run the build and test, generate assets, and send to a pipeline app to manage deployment processes. For example, Gists are a separate app, but integrate very tightly into GitHub.

I'm not sure which is the right approach, but it would be helpful to get more details.

> Also: How usable is the 0.4 branch?

Not very usable at the moment. It probably won't stabilize until end of July, although we could definitely accelerate this with more community members focused on the 0.4 branch

**Comment by bradrydzewski** [comment]
closing due to inactivity
